sky/skylet/providers/gcp/config.py

```diff
@@ -488,6 +488,93 @@ def _configure_key_pair(config, compute):
     return config
 
 
+def _check_firewall_rules(vpc_name, config, compute):
+    """Check if the firewall rules in the VPC are sufficient."""
+    required_rules = FIREWALL_RULES_TEMPLATE.copy()
+
+    operation = (
+        compute.networks().
+        getEffectiveFirewalls(
+            project=config["provider"]["project_id"],
+            network=vpc_name
+        )
+    )
+    response = operation.execute()
+    if len(response) == 0:
+        return False
+    effective_rules = response["firewalls"]
+
+    def _get_refined_rule(rule):
+        KEY_TO_COMPARE = {"sourceRanges", "allowed", "direction"}
+        refined_rule = {}
+        for k in KEY_TO_COMPARE:
+            if k not in rule:
+                continue
+            if k == "allowed":
+                refined_rule[k] = sorted(rule[k], key=lambda x: x["IPProtocol"])
+            else:
+                refined_rule[k] = rule[k]
+        return refined_rule
+
+    required_rules = list(map(_get_refined_rule, required_rules))
+    effective_rules = list(map(_get_refined_rule, effective_rules))
+
+    for rule in required_rules:
+        if rule not in effective_rules:
+            return False
+    return True
+
+
+def get_usable_vpc(config):
+    """Return a usable VPC.
+
+    If not found, create a new one with sufficient firewall rules.
+    """
+    _, _, compute, _ = construct_clients_from_provider_config(config["provider"])
+
+    # For backward compatibility, reuse the VPC if the VM is launched.
+    ins_all = _list_instances(config, compute)
+    for ins in ins_all:
+        labels = ins.get("labels", {})
+        for k, v in labels.items():
+            if k == "ray-cluster-name" and v == config["cluster_name"]:
+                if "networkInterfaces" not in ins or len(ins["networkInterfaces"]) == 0:
+                    continue
+                vpc_name = ins["networkInterfaces"][0]["network"].split("/")[-1]
+                return vpc_name
```

Comment:
Just curious, why don't we directly use the `node._list_instances()`?

Reply:
Ah yes.. I tried to use it but found that the abstraction there isn't really what we want.
The semantic of `node._list_instances()` is not listing all VMs in the zone but they put a cluster name filter
https://github.com/skypilot-org/skypilot/blob/14e26d0bd8faf9e62008f195de93b94392f8fbb8/sky/skylet/providers/gcp/node.py#L409-L411
so it will only list the VM with the associated cluster name..

I guess it's simpler that we just implement it ourselves.

Reply:
Sorry, I am confused. We also tries to filter out the clusters with a label that contains cluster_name, right?

Reply:
ah I just updated the code that uses `_list_instances()`! I misunderstood some details of the function.